egs/wsj/s5/utils/copy_data_dir.sh

```diff
@@ -116,7 +119,7 @@ fi
 if [ -f $srcdir/cmvn.scp ]; then
   utils/apply_map.pl -f 1 $destdir/spk_map <$srcdir/cmvn.scp >$destdir/cmvn.scp
 fi
-for f in stm glm ctm; do
+for f in frame_shift stm glm ctm; do
   if [ -f $srcdir/$f ]; then
     cp $srcdir/$f $destdir
```

Comment:
how about copying the conf/ dir?  Could perhaps just copy one of them?

Reply:
actually the conf dir thing is not important.  But how many different options have you tested with?

Reply:
It cannot be sensibly merged, the features can be potentially extracted differently.

> But how many different options have you tested with?

Short answer is quite exhaustively. Depends on the level of changes. For this script, basically tested the rejection if frame_shift is different, and that the file is actually copied when present somewhere (and d-oh, that initially failed, because I misspelled it). For the subset_data_dir, which I've heavily rearranged/rewritten, I had to test it exhaustively with all options (and actually used it already, but that does not cover much) -- this is a very commonly used one. Even with fake ctm files, which it also subsets, and with comments in them, because it keeps the comments; I also ran validate_data_dir on every subset it created. For validate_data_dir, I added an extra line by hand to utt2num_frames, then moved it out of order, then mangled an ints in the second columns in the third run. For simple cases where I simply had to added utt2num_frames besides utt2dir into the long list of files to be conditionally copied, basically did not do at all (do not remember if there are any changes this simple, tho).

I'm always very careful with stuff that does not have automated testing, really. Remember how I discovered the one-sample discrepancy in utt2dur extraction with and without segments the other day: I actually diffed the extracted files in the extraction scripts. All of them, and each without the segments and with a trivial segments file. Also, I've rewritten these extractions scripts such that when I did e.g. `diff make_mfcc.sh make_fbank.sh`, the only differences shown were related to the actual extractor changes. (This is where I resisted the urge to change $mfcc_dir in one and $fbank_dir in the other to $feats_dir. But maybe I should have). Then `diff make_mfcc.sh make_mfcc_pitch.sh`, and from there on `diff make_mfcc_pitch.sh make_everyother_pitch.sh`. Found some minor discrepancies this way. So... let's say I test our scrips like I would test something that has no automated test.

If only we had tests for these! If you look at the front page of Kaldi on GitHub, where it says what is the main language for the project (near the long colored horizontal bar), you'll notice that GitHub thinks Kaldi is written in bash. I was thinking about writing at least some tests, for most commonly used scripts at the very least, but this is an immense task. Just try to think of exhaustive test cases for validate_data_dir...

I quickly looked into bash testing frameworks, they'd help, but the majority of a test it still creating test data and validating it, and this is all manual. If we only head two of me...